smart-contracts/contracts/TransactionController.sol

```diff
@@ -271,4 +295,40 @@ contract TransactionController {
 
         return uint16(randomFourDigits);
     }
+
+    function withdrawFunds(uint256 _amount) external onlyRegisteredFarmers {
+        require(_amount > 0, "Amount must be greater than zero");
+        require(balances[msg.sender] >= _amount, "Insufficient funds to withdraw");
+        balances[msg.sender] = balances[msg.sender] - _amount;
+        payable(msg.sender).transfer(_amount);
+        emit FundsWithdrawn(msg.sender, _amount);
+    }
+
+    function transferFunds(address _receiver, uint256 _amount) external onlyRegisteredFarmers {
+        require(_receiver != address(0), "Address Zero not allowed");
+        require(_amount > 0, "Amount must be greater than zero");
+        require(balances[msg.sender] >= _amount, "Insufficient funds to transfer");
+        balances[msg.sender] = balances[msg.sender] - _amount;
+        payable(_receiver).transfer(_amount);
+        emit FundsTransferred(msg.sender, _receiver, _amount);
+    }
+
+    function getFarmerTotalEarnings() external view onlyRegisteredFarmers returns (uint256) {
+        return balances[msg.sender];
+    }
+
+    function getFarmerBalance() external onlyRegisteredFarmers view returns (uint256 pendingTotal, uint256 balance) {
+        address farmer = msg.sender;
+        uint256 total = 0;
+
+        for (uint i = 1; i <= transactionId; i++) {
+            Transaction storage transaction = transactions[i];
+
+            if (transaction.farmer == farmer && transaction.status == Status.Pending) {
+                total += transaction.amount;
+            }
+        }
+
+        return (total, balances[farmer]);
```

Comment:
same thing here, no need to return balances[farmer] and also, with the modifier applied, can the owner of the contract view the total farmer balance? That will be necessary for the frontend please

Reply:
same above privacy of farmers earning here, i can also add "allow owner of contract" to modifier,  The Farmer  however should be able to see their withdrawable amount, and pending amounts.. if you think its not neccessary , i can go ahead and remove it

Reply:
Allow owner of the contract to view the balances also and remove the public balances being returned